lib/pbench/server/api/resources/query_apis/search_api.py

```diff
@@ -0,0 +1,201 @@
+from flask import jsonify
+from logging import Logger
+from typing import Any, AnyStr, Dict
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources.query_apis import (
+    ElasticBase,
+    Schema,
+    Parameter,
+    ParamType,
+    PostprocessError,
+)
+
+
+class SearchAPI(ElasticBase):
+    """
+    Create a search request based on given query parameter.
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter("user", ParamType.USER, required=False),
+                Parameter("start", ParamType.DATE, required=True),
+                Parameter("end", ParamType.DATE, required=True),
+                Parameter("search_term", ParamType.STRING, required=False),
+                Parameter("fields", ParamType.LIST, required=False),
+            ),
+        )
+
+    def flatten_json(
+        self, es_json: Dict[AnyStr, Any], prefix="", out={}
+    ) -> Dict[AnyStr, Any]:
+        """
+        Utility function for flattening the json.
+        json_in =
+            {
+                    "@timestamp": "2021-07-12T14:08:46.563303",
+                    "run": {
+                        "controller": "dhcp31-171.perf.lab.eng.bos.redhat.com",
+                        "name": "pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T14.08.46"
+                    },
+                    "host_tools_info": [
+                        {
+                            "hostname": [{"test1":"dhcp31-171"}, {"test2":"dhcp31-195"}]
+                        }
+                    ]
+                }
+        json_out =
+            {
+                '@timestamp': '2021-07-12T14:08:46.563303',
+                'run.controller': 'dhcp31-171.perf.lab.eng.bos.redhat.com',
+                'run.name': 'pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T14.08.46',
+                'host_tools_info.hostname.test1': 'dhcp31-171',
+                'host_tools_info.hostname.test2': 'dhcp31-195'
+            }
+        """
+        if type(es_json) is dict:
+            for key in es_json:
+                while prefix[-2:-1] == ".":
+                    prefix = prefix[:-1]
+                self.flatten_json(es_json[key], prefix + key + ".", out)
+        elif type(es_json) is list:
+            i = 0
+            for element in es_json:
+                if type(element) is dict:
+                    self.flatten_json(element, prefix + ".", out)
+                else:
+                    self.flatten_json(element, prefix + str(i) + ".", out)
+                i += 1
+        else:
+            out[prefix[:-1]] = es_json
```

Comment:
Do we really want `{"": 1}` as a transformation for the "JSON" value of `1`? Not that this is likely to happen for any `/_search` response, but if the code handles it ...

Reply:
By JSON value of 1 you mean `{1}` or `{"": 1}`?

Reply:
As I said, I was just thinking through edge cases. If the Elasticsearch `["_source"]` for a "hit" were to have the value `1`, then `postprocess` will pass that degenerate JSON to `flatten_json`, which will return `{"": 1}`.

I don't think we really need to worry about this much, as it would be a bizarre Elasticsearch behavior, and I doubt it's even possible.

Reply:
I haven't gotten far enough to review this code, yet, but I would point out that `{1}` is not a valid JSON object.  (However, `{"":1}`, much to my shock, *is*...and it's a valid Python `dict` too...I had no idea that the empty string was a valid key.)

I think the first question is, is the input and/or output of this flattening supposed to be an _object_?  If so, then perhaps we can cleave this Gordian knot and not worry about scalar values at the top level (or render them as an error).

Reply:
Yes, I ran through the same experiment. Heck, I was surprised that `""[:-1]` works and returns an empty string, although perhaps I shouldn't have been.

The input ought to be the Elasticsearch template document, which should be a JSON object; so we would never expect either `1` or `{1}` at the top level. Still, "what if" is always a fun game.